docs/docsite/rst/dev_guide/overview_architecture.rst

```diff
@@ -10,26 +10,34 @@ It uses no agents and no additional custom security infrastructure, so it's easy
 
 In this section, we'll give you a really quick overview of how Ansible works so you can see how the pieces fit together.
 
+.. contents::
+   :local:
+
 Modules
 =======
 
-Ansible works by connecting to your nodes and pushing out small programs, called "Ansible Modules" to them. These programs are written to be resource models of the desired state of the system. Ansible then executes these modules (over SSH by default), and removes them when finished.
+Ansible works by connecting to your nodes and pushing out small programs, called "Ansible Modules" to them. These programs are written to be resource models of the desired state of the system. Ansible then executes these modules (over SSH by default), and removes them when finished. Your library of modules can reside on any machine, and there are no servers, daemons, or databases required.
+
+You can :ref:`write your own modules <developing_modules_general>`, though you should first consider :ref:`whether you should <developing_modules>`. Typically you'll work with your favorite terminal program, a text editor, and probably a version control system to keep track of changes to your content.
+
+Module utilities
+================
 
-Your library of modules can reside on any machine, and there are no servers, daemons, or databases required. Typically you'll work with your favorite terminal program, a text editor, and probably a version control system to keep track of changes to your content.
+When multiple modules use the same code, Ansible stores those functions as module utilities. For example, the code that parses URLs is ``lib/ansible/module_utils/url.py``. You can :ref:`write your own module utilities <appendix_module_utilities>` as well.
 
 Plugins
 =======
 
-Plugins are pieces of code that augment Ansible's core functionality. Ansible ships with a number of handy plugins, and you can easily write your own.
+Plugins augment Ansible's core functionality with features like cacheing, callbacks, filters, and lookups. Plugins are accessible to all modules. Ansible ships with a number of handy plugins, and you can easily :ref:`write your own <developing_plugins>`.
```

Comment:
The list of features seems more like a list of plugin types.  Maybe something more along the lines of "Ansible's core functionality can be extended by a variety of plugin types".

I'm not quite sure what "Plugins are accessible to all modules" means?

Reply:
Yeah, this paragraph needs work. How would you describe the difference between a plugin and a module utility? Both are "shared code" in one sense, but they are not the same thing - that's what I was trying to capture here.

Reply:
The new wording looks good to me. 

plugins are very similar to modules.  The notable differences being:
* plugins execute on the controller versus modules on the managed machine
* plugins are run inside of the /usr/bin/ansible process while modules are run in a separate process.

module_utils are different.  They're libraries of code that provide building blocks for implementing other things (modules, plugins, and even core code).

In the old code, it was difficult for one plugin to reuse code from another plugin (nitzmahone would know whether the new code makes that possible or not).  So plugins being accessible might not be correct?

Reply:
good info, thanks - I've incorporated this in the changeset